### PR TITLE
bugfix/unregistration memfile event handling

### PR DESCRIPTION
**Pull request by rex-schilasky**
### Description
The new introduced "active register/unregister logic" calls depending on the connection type specific functions like (AddLocConnection/RemLocConnection,AddExtConnection/RemExtConnection) for the SHM and UDP data writer (TCP was just not handled for now).

The RemLocConnection was not called in that context before the new unregister/register logic and is closing the 2 connection events for updating and acknowledging memory file data transfer/receive. This leads to a dead shm connection when the subscriber on the other side is destroyed (triggering RemLocConnection in the publishing process) and created again.

Solution: Events are not closed on that level but are cleaned up together with the memory file resource (CSyncMemFile) properly.

### Related issues
No issue now, will be created soon ..

### Cherry-pick to
- None

### CODE REVIEW

File: ecal/core/src/readwrite/ecal_writer_base.h [hunk 1]
```diff
@@ -48,11 +48,11 @@ namespace eCAL
...
 
-    virtual bool AddLocConnection(const std::string& /*process_id_*/, const std::string& /*conn_par_*/) { return false; };
-    virtual bool RemLocConnection(const std::string& /*process_id_*/) { return false; };
+    virtual bool AddLocConnection(const std::string& /*process_id_*/, const std::string& /*topic_id_*/, const std::string& /*conn_par_*/) { return false; };
+    virtual bool RemLocConnection(const std::string& /*process_id_*/, const std::string& /*topic_id_*/) { return false; };
 
-    virtual bool AddExtConnection(const std::string& /*host_name_*/, const std::string& /*process_id_*/, const std::string& /*conn_par_*/) { return false; };
-    virtual bool RemExtConnection(const std::string& /*host_name_*/, const std::string& /*process_id_*/) { return false; };
+    virtual bool AddExtConnection(const std::string& /*host_name_*/, const std::string& /*process_id_*/, const std::string& /*topic_id_*/, const std::string& /*conn_par_*/) { return false; };
+    virtual bool RemExtConnection(const std::string& /*host_name_*/, const std::string& /*process_id_*/, const std::string& /*topic_id_*/) { return false; };
```
Comment (KerstinKeller) [inline ecal/core/src/readwrite/ecal_writer_base.h:56]: Why did you change the function signatures? Are the arguments actually being used somewhere?

File: ecal/core/src/readwrite/ecal_writer.cpp [hunk 2]
```diff
@@ -764,22 +772,15 @@ namespace eCAL
...
 
       m_loc_subscribed = !m_loc_sub_map.empty();
       m_ext_subscribed = !m_ext_sub_map.empty();
     }
 
-    for(const auto& loc_sub : *loc_timeouts)
-    {
-      m_writer.shm.RemLocConnection(loc_sub.process_id);
-    }
-
```
Comment (KerstinKeller) [inline ecal/core/src/readwrite/ecal_writer.cpp:782]: This is not the "goodbye" message but happens on registration timeout?

File: ecal/core/src/readwrite/ecal_writer_shm.cpp [hunk 3]
```diff
@@ -190,28 +190,6 @@ namespace eCAL
...
-      // TODO: Fix this in 'CSyncMemoryFile::Disconnect' to handle event resources properly.
-      if (std::to_string(eCAL::Process::GetProcessID()) != process_id_)
-      {
-        ret_state &= memory_file->Disconnect(process_id_);
-      }
-    }
-
-    return ret_state;
-  }
-
```
Comment (KerstinKeller) [inline ecal/core/src/readwrite/ecal_writer_shm.cpp:214]: It feels a bit wrong that connections cannot be removed again. So the resources stay open until the publisher is closed? Is that what happens? Also they continue being signalled, etc?

And btw, I guess you have addd line 206 to prevent the testcase to fail. see the huge comment. Maybe it was just not working at all on Posix...

File: ecal/core/src/readwrite/ecal_writer.cpp [hunk 4]
```diff
@@ -686,8 +688,10 @@ namespace eCAL
     }
 
     // remove a local subscription
-    m_writer.udp_mc.RemLocConnection (local_info_.process_id);
-    m_writer.shm.RemLocConnection    (local_info_.process_id);
+    // currently all these functions are not implemented (there is no need to react on)
+    m_writer.udp_mc.RemLocConnection (local_info_.process_id, local_info_.topic_id);
+    m_writer.shm.RemLocConnection    (local_info_.process_id, local_info_.topic_id);
+    m_writer.tcp.RemLocConnection    (local_info_.process_id, local_info_.topic_id);
```
Comment (KerstinKeller) [inline ecal/core/src/readwrite/ecal_writer.cpp:695]: Adding the tcp `RemLocConnection` / `AddExtConnection` should preferably be done in a different PR as it is not related to the issue that is being fixed here.

**Review by KerstinKeller — COMMENTED** [review]
For me, a lot of questions (concerning) the original design, still remain.

I find it a bit questionable, to keep all signalling resources open until the memory file itself is destroyed.
Couldn't it lead to the same problems on subscriber side, e.g. if the subscriber side destroys the events (e..g after 60 second timeout) and then creates a new subscriber?
(E.g. now we made sure events are not closed on sender side, because in the testcase they stayed open on subscriber side. What if they are closed/reopened on subscriber side... in that case communication can't be established, either?)

It feels to me, like we might have just shifted the problem.
Maybe it's worth to take a look at the underlying NamedEvents.

**Review by KerstinKeller — CHANGES_REQUESTED** [review]
Now it doesn't work in the case that the subscriber is calling `finalize()` and `initialize()`

File: ecal/core/src/readwrite/ecal_writer_base.h (hunk 1 above)
Comment (rex-schilasky) [reply ecal/core/src/readwrite/ecal_writer_base.h:56]: They are not used yet, but IF any layer wants to intercept that information it makes sense to provide host_name, process_id, topic_id to identify exactly the connection the is added or removed. The old signature supported just SHM with the needed information process_id.

File: ecal/core/src/readwrite/ecal_writer.cpp (hunk 2 above)
Comment (rex-schilasky) [reply ecal/core/src/readwrite/ecal_writer.cpp:782]: Yes it's not the goodbye message and no more needed at all. It closed the connected events for the memory file that was connected to this shm writer. This is now done since many version in the destruction of the CSyncMemFile.

File: ecal/core/src/readwrite/ecal_writer_shm.cpp (hunk 3 above)
Comment (rex-schilasky) [reply ecal/core/src/readwrite/ecal_writer_shm.cpp:214]: No it's no more signalling, it is just holding the resources for MemFile + Events. The publisher (writer) knows the number of connections local/extern all the time.

File: ecal/core/src/readwrite/ecal_writer.cpp (hunk 4 above)
Comment (rex-schilasky) [reply ecal/core/src/readwrite/ecal_writer.cpp:695]: I will mark this PR as a cleanup and we do the fix in another one.

**Comment by rex-schilasky** [comment]
> Now it doesn't work in the case that the subscriber is calling `finalize()` and `initialize()`

Yes, and I assume that this issue is existing independent from that "fix". I will check this again.

**Comment by rex-schilasky** [comment]
mixed up bugfixes and code cleanups + did not really fix the issue
